### DISCUSSION TITLE
adventures with plugin-transform-typescript and allowDeclareFields

### DISCUSSION THREAD

**Discussion by gurgeous** [babel/babel#12534] [Q&A] [answered]
Thanks so much for creating babel, it's hugely helpful to our project!

We recently switched to `@babel/plugin-transform-typescript` @ 7.12.1. Previously we were using the typescript compiler directly via webpack loader. Unfortunately, vast swaths of our app stopped working due to the behavior of `allowDeclareFields` / `declare``. Here is my attempt to summarize the situation. Please forgive my ignorance, I'm not an expert on many of the packages involved here.

1. We use typescript & vue with class components. After the change, all of our vue components broke because `$refs` was getting set to `void 0`. Vue crashes when `$refs` is set to undefined. You can sorta fix this manually by assigning a value to $refs like `{}`, but that would require a ts workaround since it violates the types we've setup for $refs.
1. After much digging, we decided to set `allowDeclareFields: true` and add `declare` to $refs. We were pointed toward this solution by the babel docs, which state that this will be the default in babel 8. This seemed to work, though it required a lengthy set of changes to our babel config because we couldn't get `allowDeclareFields` to work directly with `babel-preset-typescript-vue`.
1. Later, we realized that other parts of our app were also broken in subtle ways due to fields incorrectly getting set to undefined. For example, we have an abstract Model class that assigns fields from JSON in the constructor. With the new behavior in babel, subclasses of Model will unfortunately set all those fields to `undefined` after the Model constructor has run. That overwrites all the JSON data we just set, whoops.
1. After more consideration, we decided we lacked the confidence to correctly use `allowDeclareFields: true`. We were unable to determine the purpose of `declare`, the pros and cons of the behavior, and where it might cause problems with our project.
1. We set allowDeclareFields: false to disable this feature.

It's also confusing that there seem to be three states for allowDeclareFields. Does this look correct?

1. Not set. Fields like `$refs` ARE set to undefined and `declare` is not allowed. (this is what we blundered into accidentally)
1. Set to true. Fields ARE set to undefined and `declare` is allowed. (we tried this and abandoned)
1. Set to false. Fields are NOT set to undefined and `declare` is not allowed. (our final fix)

Here is a simple Vue component that will crash:

```ts
@Component({})
export default class Something extends Vue {
  public $refs: {
    div: HTMLDivElement;
  };
 ...
}
```

**Comment by nicolo-ribaudo** [comment, marked as answer]
_(I'm converting this issue to a discussion)_

When TypeScript first implemented class properties, the current [class fields proposal](https://github.com/tc39/proposal-class-fields/) was still far from the current status. For this reason, the TypeScript team had to make some design decisions about them.

The problem is that TypeScript's implementation of class fields was incompatible with the proposal, in two different ways:
1. TypeScript _sets_ the fields, instead of _defining_ them: it uses `this.foo = ...` instead of `Object.defineProperty(this, "foo", ...)` (https://github.com/microsoft/TypeScript/issues/34787)
2. TypeScript _ignores_ class fields without an initializer, instead of initializing them to `undefined`. (https://github.com/microsoft/TypeScript/issues/12437)

Since fixing those two TypeScript differences* from the proposal would make it impossible to declare class fields just for their type annotation, the TypeScript team introduced the `declare` keyword.

**When do you need `declare`?** If you have a class field like `class X { foo: number; }` and you _don't want_ it to be the same as `class X { foo: number = undefined; }`, then you need to use `class X { declare foo: number; }`.

Even if you don't decide to migrate _now_ to `declare`, please keep in mind that:
- In Babel 8, `allowDeclareFields` will be enabled by default and it will not be possible to disable it
- In a future TypeScript version, TypeScript will change its implementation to match the spec (and Babel's implementation with `allowDeclareFields: true`)

\* rather than differences we could also call them "bugs" (even if there is a very strong historical reason for them), because TypeScript's goal is to closely match ECMAScript's semantics.

---

> It's also confusing that there seem to be three states for allowDeclareFields. Does this look correct?

No, `false` and "not set" should have the same behavior. I think that the difference in the behavior you are seeing in (1) and in (3) is caused by changing the order that the TypeScript plugin and the class fields plugin run.

Since the class fields plugin matches the spec, if it runs _before_ the TypeScript one you will already get the spec compliant behavior.
In (3) probably the TypeScript plugin is running first, this exposing the current default TS behavior.

**Comment by gurgeous** [comment]
Thank you for taking the time to write such a great response, @nicolo-ribaudo! Hopefully this issue might help some other folks avoid days of sleuthing to understand the new behavior.

One humble suggestion - we should beef up the docs on this topic, perhaps pointing readers at this very discussion.
